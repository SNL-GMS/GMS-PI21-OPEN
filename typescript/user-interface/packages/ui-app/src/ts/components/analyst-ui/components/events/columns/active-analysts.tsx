import type {
  CellRendererParams,
  ColumnDefinition,
  TooltipParams,
  ValueFormatterParams,
  ValueGetterParams
} from '@gms/ui-core-components';
import { EventsColumn } from '@gms/ui-state';
import { getTextWidth } from '@gms/ui-util';
import type Immutable from 'immutable';

import { medCellWidthPx } from '~common-ui/common/table-types';

import type { EventRow } from '../types';
import { columnDisplayStrings } from '../types';

/**
 * Column value getter params - value is of type string
 */
export type EventValueGetterParams = ValueGetterParams<
  EventRow,
  unknown,
  string,
  CellRendererParams<any, any, any, any, any>,
  unknown
>;
/**
 * Returns the value getter for the active analysts array.
 *
 * @param params the table data params
 * @returns the formatted value
 */
export const activeAnalystsValueGetter = (params: EventValueGetterParams): string =>
  params.data?.activeAnalysts?.join(', ');

/**
 * Returns the formatted value for the active analysts array.
 *
 * @param params the table data params
 * @returns the formatted value
 */
export const activeAnalystsValueFormatter = (
  params: ValueFormatterParams<
    EventRow,
    unknown,
    string,
    CellRendererParams<any, any, any, any, any>,
    unknown
  >
): string => {
  let names = '';
  if (params !== undefined && params.data !== undefined) {
    const {
      data: { activeAnalysts },
      column
    } = params;
    const PADDING = 65;
    const colWidth = column ? column.getActualWidth() : Number.MAX_SAFE_INTEGER;

    // eslint-disable-next-line no-plusplus
    for (let i = 1; activeAnalysts && i <= activeAnalysts.length; i++) {
      names = activeAnalysts.slice(0, i).join(', ');
      const textWidth = Number(getTextWidth(names)) + PADDING;
      if (textWidth >= colWidth) {
        names = `${activeAnalysts.slice(0, Math.max(i - 1, 1)).join(', ')} +
       ${Math.min(activeAnalysts.length - 1, activeAnalysts.length + 1 - i)}`;
        break;
      }
    }
  }

  return names;
};

/**
 * Returns the tooltip value getter for the active analysts.
 */
export const activeAnalystsTooltipValueGetter = (params: TooltipParams): string => params.value;

/**
 * Defines the active analysts column definition
 *
 * @param columnsToDisplayMap a map that specifies if the column is visible
 * @returns the column definition
 */
export const activeAnalystsColumnDef = (
  columnsToDisplayMap: Immutable.Map<EventsColumn, boolean>
): ColumnDefinition<EventRow, unknown, string, unknown, unknown> => ({
  headerName: columnDisplayStrings.get(EventsColumn.activeAnalysts),
  field: EventsColumn.activeAnalysts,
  headerTooltip: columnDisplayStrings.get(EventsColumn.activeAnalysts),
  width: medCellWidthPx,
  hide: !columnsToDisplayMap.get(EventsColumn.activeAnalysts),
  valueGetter: activeAnalystsValueGetter,
  valueFormatter: activeAnalystsValueFormatter,
  tooltipValueGetter: activeAnalystsTooltipValueGetter
});
