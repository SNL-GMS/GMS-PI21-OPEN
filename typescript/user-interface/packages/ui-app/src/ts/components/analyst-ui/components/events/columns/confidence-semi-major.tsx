import { setDecimalPrecision } from '@gms/common-util';
import type { ColumnDefinition } from '@gms/ui-core-components';
import { EventsColumn } from '@gms/ui-state';
import type Immutable from 'immutable';

import type { EventRow } from '../types';
import { columnDisplayStrings } from '../types';

/**
 * Defines the confidence semi major column definition
 *
 * @param columnsToDisplayMap a map that specifies if the column is visible
 * @returns the column definition
 */
export const confidenceSemiMajorColumnDef = (
  columnsToDisplayMap: Immutable.Map<EventsColumn, boolean>
): ColumnDefinition<EventRow, unknown, number, unknown, unknown> => ({
  headerName: columnDisplayStrings.get(EventsColumn.confidenceSemiMajorAxis),
  field: EventsColumn.confidenceSemiMajorAxis,
  headerTooltip: 'Confidence semi-major axis in kilometers',
  width: 130,
  hide: !columnsToDisplayMap.get(EventsColumn.confidenceSemiMajorAxis),
  valueFormatter: params => setDecimalPrecision(params.data.confidenceSemiMajorAxis, 2),
  filterValueGetter: params => setDecimalPrecision(params.data.confidenceSemiMajorAxis, 2)
});
