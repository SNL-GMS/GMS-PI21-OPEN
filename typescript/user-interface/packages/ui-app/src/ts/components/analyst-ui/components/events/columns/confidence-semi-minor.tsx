import { setDecimalPrecision } from '@gms/common-util';
import type { ColumnDefinition } from '@gms/ui-core-components';
import { EventsColumn } from '@gms/ui-state';
import type Immutable from 'immutable';

import type { EventRow } from '../types';
import { columnDisplayStrings } from '../types';

/**
 * Defines the confidence semi minor column definition
 *
 * @param columnsToDisplayMap a map that specifies if the column is visible
 * @returns the column definition
 */
export const confidenceSemiMinorColumnDef = (
  columnsToDisplayMap: Immutable.Map<EventsColumn, boolean>
): ColumnDefinition<EventRow, unknown, number, unknown, unknown> => ({
  headerName: columnDisplayStrings.get(EventsColumn.confidenceSemiMinorAxis),
  field: EventsColumn.confidenceSemiMinorAxis,
  headerTooltip: 'Confidence semi-minor axis in kilometers',
  width: 130,
  hide: !columnsToDisplayMap.get(EventsColumn.confidenceSemiMinorAxis),
  valueFormatter: params => setDecimalPrecision(params.data.confidenceSemiMinorAxis, 2),
  filterValueGetter: params => setDecimalPrecision(params.data.confidenceSemiMinorAxis, 2)
});
