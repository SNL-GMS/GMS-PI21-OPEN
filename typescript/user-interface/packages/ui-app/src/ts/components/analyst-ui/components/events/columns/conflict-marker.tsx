import { Icon, IconSize, Intent } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import type { CellRendererParams, ColumnDefinition } from '@gms/ui-core-components';
import { EventsColumn, useAppSelector } from '@gms/ui-state';
import classNames from 'classnames';
import type Immutable from 'immutable';
import * as React from 'react';

import type { EventRow } from '../types';
import { columnDisplayStrings } from '../types';

/**
 * Cell renderer to render the conflict marker column
 */
export const ConflictMarkerCellRenderer: React.FunctionComponent<CellRendererParams<
  EventRow,
  unknown,
  boolean,
  unknown,
  unknown
  // eslint-disable-next-line react/function-component-definition
>> = (props: CellRendererParams<EventRow, unknown, boolean, unknown, unknown>) => {
  const openEventId = useAppSelector(state => state.app.analyst.openEventId);
  const { data } = props;
  const color = data.id === openEventId ? 'black' : 'red';
  const isConflict = data.conflict;
  return !isConflict ? (
    <div className={classNames('table-cell', 'events__conflict-marker')}>
      <Icon icon={IconNames.ISSUE} intent={Intent.DANGER} size={IconSize.LARGE} color={color} />
    </div>
  ) : (
    <div className="table-cell" style={{ height: '36px' }}>
      <span>TBD</span>
    </div>
  );
};

/**
 * Defines the conflict marker column definition
 *
 * @param columnsToDisplayMap a map that specifies if the column is visible
 * @returns the column definition
 */
export const conflictColumnDef = (
  columnsToDisplayMap: Immutable.Map<EventsColumn, boolean>
): ColumnDefinition<EventRow, unknown, boolean, unknown, unknown> => ({
  headerName: columnDisplayStrings.get(EventsColumn.conflict),
  field: EventsColumn.conflict,
  headerTooltip: columnDisplayStrings.get(EventsColumn.conflict),
  width: 100,
  hide: !columnsToDisplayMap.get(EventsColumn.conflict),
  cellRendererFramework: ConflictMarkerCellRenderer,
  pinned: 'left'
});
