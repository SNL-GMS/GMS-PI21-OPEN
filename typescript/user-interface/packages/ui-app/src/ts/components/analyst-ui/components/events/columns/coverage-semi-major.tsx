import { setDecimalPrecision } from '@gms/common-util';
import type { ColumnDefinition } from '@gms/ui-core-components';
import { EventsColumn } from '@gms/ui-state';
import type Immutable from 'immutable';

import type { EventRow } from '../types';
import { columnDisplayStrings } from '../types';

/**
 * Defines the coverage semi major column definition
 *
 * @param columnsToDisplayMap a map that specifies if the column is visible
 * @returns the column definition
 */
export const coverageSemiMajorColumnDef = (
  columnsToDisplayMap: Immutable.Map<EventsColumn, boolean>
): ColumnDefinition<EventRow, unknown, number, unknown, unknown> => ({
  headerName: columnDisplayStrings.get(EventsColumn.coverageSemiMajorAxis),
  field: EventsColumn.coverageSemiMajorAxis,
  headerTooltip: 'Coverage semi-major axis in kilometers',
  width: 130,
  hide: !columnsToDisplayMap.get(EventsColumn.coverageSemiMajorAxis),
  valueFormatter: params => setDecimalPrecision(params.data.coverageSemiMajorAxis, 2),
  filterValueGetter: params => setDecimalPrecision(params.data.coverageSemiMajorAxis, 2)
});
