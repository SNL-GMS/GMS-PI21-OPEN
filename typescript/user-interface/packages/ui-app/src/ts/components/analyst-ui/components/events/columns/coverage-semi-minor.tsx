import { setDecimalPrecision } from '@gms/common-util';
import type { ColumnDefinition } from '@gms/ui-core-components';
import { EventsColumn } from '@gms/ui-state';
import type Immutable from 'immutable';

import type { EventRow } from '../types';
import { columnDisplayStrings } from '../types';

/**
 * Defines the coverage semi minor column definition
 *
 * @param columnsToDisplayMap a map that specifies if the column is visible
 * @returns the column definition
 */
export const coverageSemiMinorColumnDef = (
  columnsToDisplayMap: Immutable.Map<EventsColumn, boolean>
): ColumnDefinition<EventRow, unknown, number, unknown, unknown> => ({
  headerName: columnDisplayStrings.get(EventsColumn.coverageSemiMinorAxis),
  field: EventsColumn.coverageSemiMinorAxis,
  headerTooltip: 'Coverage semi-minor axis in kilometers',
  width: 130,
  hide: !columnsToDisplayMap.get(EventsColumn.coverageSemiMinorAxis),
  valueFormatter: params => setDecimalPrecision(params.data.coverageSemiMinorAxis, 2),
  filterValueGetter: params => setDecimalPrecision(params.data.coverageSemiMinorAxis, 2)
});
