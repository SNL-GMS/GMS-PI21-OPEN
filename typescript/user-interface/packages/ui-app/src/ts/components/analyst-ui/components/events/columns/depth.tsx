import { setDecimalPrecision } from '@gms/common-util';
import type { ColumnDefinition } from '@gms/ui-core-components';
import { EventsColumn } from '@gms/ui-state';
import type Immutable from 'immutable';

import type { EventRow } from '../types';
import { columnDisplayStrings } from '../types';

/**
 * Defines the depth column definition
 *
 * @param columnsToDisplayMap a map that specifies if the column is visible
 * @returns the column definition
 */
export const depthColumnDef = (
  columnsToDisplayMap: Immutable.Map<EventsColumn, boolean>
): ColumnDefinition<EventRow, unknown, number, unknown, unknown> => ({
  headerName: columnDisplayStrings.get(EventsColumn.depthKm),
  field: EventsColumn.depthKm,
  headerTooltip: 'Depth in kilometers',
  hide: !columnsToDisplayMap.get(EventsColumn.depthKm),
  valueFormatter: params => setDecimalPrecision(params.data.depthKm, 3),
  filterValueGetter: params => setDecimalPrecision(params.data.depthKm, 3)
});
