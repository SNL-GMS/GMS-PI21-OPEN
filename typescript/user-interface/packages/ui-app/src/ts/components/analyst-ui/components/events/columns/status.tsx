import { EventTypes } from '@gms/common-model';
import { humanReadable, toSentenceCase } from '@gms/common-util';
import type { CellRendererParams, ColumnDefinition } from '@gms/ui-core-components';
import { getColumnPosition, TableCellRenderer } from '@gms/ui-core-components';
import { EventsColumn } from '@gms/ui-state';
import type Immutable from 'immutable';
import * as React from 'react';

import type { EventRow } from '../types';
import { columnDisplayStrings } from '../types';

/**
 * Cell renderer to render the status column
 */
export function StatusCellRenderer(
  props: CellRendererParams<EventRow, unknown, string, unknown, unknown>
) {
  const { data } = props;
  const value = data.status
    ? toSentenceCase(humanReadable(data.status)) ??
      toSentenceCase(humanReadable(EventTypes.EventStatus.NOT_STARTED))
    : toSentenceCase(humanReadable(EventTypes.EventStatus.NOT_STARTED));
  return (
    <TableCellRenderer
      data-col-position={getColumnPosition<EventRow>(props)}
      value={value}
      isNumeric={false}
      tooltipMsg={value}
      cellValueClassName={
        data.status === EventTypes.EventStatus.COMPLETE ? 'table-cell__value--complete' : ''
      }
    />
  );
}

/**
 * Defines the status column definition
 *
 * @param columnsToDisplayMap a map that specifies if the column is visible
 * @returns the column definition
 */
export const statusColumnDef = (
  columnsToDisplayMap: Immutable.Map<EventsColumn, boolean>
): ColumnDefinition<EventRow, unknown, string, unknown, unknown> => ({
  headerName: columnDisplayStrings.get(EventsColumn.status),
  field: EventsColumn.status,
  headerTooltip: columnDisplayStrings.get(EventsColumn.status),
  width: 110,
  hide: !columnsToDisplayMap.get(EventsColumn.status),
  pinned: 'right',
  cellRendererFramework: StatusCellRenderer
});
