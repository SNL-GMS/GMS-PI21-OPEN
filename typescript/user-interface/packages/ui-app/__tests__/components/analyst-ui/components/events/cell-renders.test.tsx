/* eslint-disable @typescript-eslint/no-unused-vars */
import { analystActions, getStore } from '@gms/ui-state';
import { UILogger } from '@gms/ui-util';
import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';

import { ConflictMarkerCellRenderer } from '../../../../../src/ts/components/analyst-ui/components/events/columns/conflict-marker';
import { StatusCellRenderer } from '../../../../../src/ts/components/analyst-ui/components/events/columns/status';
import type { EventTableCellRendererParams } from '../../../../../src/ts/components/analyst-ui/components/events/types';
import { dummyData, dummyData2 } from './event-table-mock-data';

const logger = UILogger.create('GMS_CELL_RENDERS_TEST', process.env.GMS_CELL_RENDERS_TEST);

const props1: EventTableCellRendererParams<boolean> = {
  api: undefined,
  colDef: undefined,
  column: undefined,
  columnApi: undefined,
  context: undefined,
  data: dummyData2,
  eGridCell: undefined,
  eParentOfValue: undefined,
  formatValue: undefined,
  getValue: undefined,
  node: undefined,
  refreshCell: undefined,
  rowIndex: undefined,
  setValue: undefined,
  value: true,
  valueFormatted: undefined,
  registerRowDragger(
    rowDraggerElement: HTMLElement,
    dragStartPixels?: number,
    value?: string
  ): void {
    logger.info('registerRowDragger function not implemented');
  }
};

const props2: EventTableCellRendererParams<string> = {
  api: undefined,
  colDef: undefined,
  column: undefined,
  columnApi: undefined,
  context: undefined,
  data: dummyData2,
  eGridCell: undefined,
  eParentOfValue: undefined,
  formatValue: undefined,
  getValue: undefined,
  node: undefined,
  refreshCell: undefined,
  rowIndex: undefined,
  setValue: undefined,
  value: undefined,
  valueFormatted: undefined,
  registerRowDragger(
    rowDraggerElement: HTMLElement,
    dragStartPixels?: number,
    value?: string
  ): void {
    logger.info('registerRowDragger function not implemented');
  }
};

describe('ConflictMarkerCellRenderer', () => {
  it('is exporter', () => {
    expect(ConflictMarkerCellRenderer).toBeDefined();
  });

  it('matches the snapshot', () => {
    const store = getStore();
    store.dispatch(analystActions.setOpenEventId('test'));

    // snapshot with conflict
    let renderResult = render(
      <Provider store={store}>
        {/* eslint-disable-next-line react/jsx-props-no-spreading */}
        <ConflictMarkerCellRenderer {...props1} />
      </Provider>
    );
    expect(renderResult.container).toMatchSnapshot();

    props1.data = dummyData;
    store.dispatch(analystActions.setOpenEventId(dummyData.id));

    // snapshot without conflict
    // eslint-disable-next-line react/jsx-props-no-spreading
    renderResult = render(
      <Provider store={store}>
        {/* eslint-disable-next-line react/jsx-props-no-spreading */}
        <ConflictMarkerCellRenderer {...props1} />
      </Provider>
    );
    expect(renderResult.container).toMatchSnapshot();
  });
});

describe('StatusCellRenderer', () => {
  it('is exporter', () => {
    expect(StatusCellRenderer).toBeDefined();
  });

  it('matches the snapshot', () => {
    props2.data = dummyData2;
    // snapshot with complete status
    // eslint-disable-next-line react/jsx-props-no-spreading
    let renderResult = render(<StatusCellRenderer {...props2} />);
    expect(renderResult.container).toMatchSnapshot();

    props2.data = dummyData;
    // snapshot without different status
    // eslint-disable-next-line react/jsx-props-no-spreading
    renderResult = render(<StatusCellRenderer {...props2} />);
    expect(renderResult.container).toMatchSnapshot();
  });
});
