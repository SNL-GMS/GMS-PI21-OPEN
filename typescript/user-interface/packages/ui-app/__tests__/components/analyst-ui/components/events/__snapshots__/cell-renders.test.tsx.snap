// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConflictMarkerCellRenderer matches the snapshot 1`] = `
<div>
  <div
    class="table-cell events__conflict-marker"
  >
    <span
      aria-hidden="true"
      class="bp4-icon bp4-icon-issue bp4-intent-danger"
      icon="issue"
    >
      <svg
        data-icon="issue"
        fill="red"
        height="20"
        viewBox="0 0 20 20"
        width="20"
      >
        <path
          d="M10 20C4.477 20 0 15.523 0 10S4.477 0 10 0s10 4.477 10 10-4.477 10-10 10zm0-2a8 8 0 100-16 8 8 0 000 16zm1-2H9v-2h2v2zm0-3H9V4h2v9z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`ConflictMarkerCellRenderer matches the snapshot 2`] = `
<div>
  <div
    class="table-cell"
    style="height: 36px;"
  >
    <span>
      TBD
    </span>
  </div>
</div>
`;

exports[`StatusCellRenderer matches the snapshot 1`] = `
<div>
  <div
    class="table-cell "
    data-cy="table-cell"
    style="height: 36px; overflow: hidden;"
    title="Complete"
  >
    <div
      class=" table-cell__value table-cell__value--complete"
      data-cy="table-cell__value"
    >
      <span>
        Complete
      </span>
    </div>
  </div>
</div>
`;

exports[`StatusCellRenderer matches the snapshot 2`] = `
<div>
  <div
    class="table-cell "
    data-cy="table-cell"
    style="height: 36px; overflow: hidden;"
    title="In progress"
  >
    <div
      class=" table-cell__value "
      data-cy="table-cell__value"
    >
      <span>
        In progress
      </span>
    </div>
  </div>
</div>
`;
